### PR TITLE
Fix a bug in reconciliation.

### PR DESCRIPTION
**Pull request by agorrod**
In cases where we decide to squeeze one more large entry onto a page,
we could have attempted to decrement space available negative.

### CODE REVIEW

**Comment by agorrod** [comment]
@keithbostic I ran into a case where the assertion at ```rec_write.c:1993``` was firing.

After some debugging it was because we are electing to take the ```__rec_split_grow``` path the first time through, when there are 14 entries on the page and one more entry pushes over a page boundary. That is the first 14 entries take 1173 bytes, the 15th entry would be 3000 bytes, the page size is configured to be 4096, and the ```leaf_value_max``` is 3096.

It's difficult to get a database in the right state to reproduce this one, but once I did it was relatively simple to trigger the failure. So I've validated that this change works.

I tried another change which also fixed the problem:

```
--- a/src/reconcile/rec_write.c
+++ b/src/reconcile/rec_write.c
@@ -1951,7 +1951,8 @@ __rec_split(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t next_len)
                 * we haven't already consumed a reasonable portion of a split
                 * chunk.
                 */
-               if (inuse < r->split_size / 2)
+               if (inuse < r->split_size / 2 &&
+                   (r->entries == 1 || inuse + next_len < r->page_size))
                        break;
 
                /*
```

The difference between the change in the pull request and the change in the above diff is. In the pull request we'll put the large item onto the page (even though it'll mean the page is larger than the configured page size), in the change in the diff we'll put the large item on a page by itself.

File: src/reconcile/rec_write.c [hunk 1]
```diff
@@ -1982,14 +1982,18 @@ __rec_split(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t next_len)
...
+
+		/*
+		 * If there is room but not enough room for a split size chunk
+		 * try to fit the rest of the content onto the page.
+		 * We may have already gathered more than a page worth of
+		 * data if there has been a large value.
+		 */
+		if (inuse <= r->page_size &&
+		    inuse + r->space_avail > r->page_size) {
 			WT_ASSERT(session, r->page_size >= inuse);
```
Comment (keithbostic) [inline src/reconcile/rec_write.c]: Thanks, @agorrod, nice catch.

My concern about this change is we don't reset `r->space_avail`, it's still set to the next split size, so if `inuse` is greater than page size, but `next_len` is less than the split size, we'll keep on adding to the chunk.

What would you think of this change, instead?

The idea is that if we've grown bigger than a page, there's no space available, and we'll fall through into the `SPLIT_MAX` case immediately.

```
diff --git a/src/reconcile/rec_write.c b/src/reconcile/rec_write.c
index 4ca1a6b..cceaf71 100644
--- a/src/reconcile/rec_write.c
+++ b/src/reconcile/rec_write.c
@@ -1985,13 +1985,15 @@ __rec_split(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t next_len)
                /*
                 * Set the space available to another split-size chunk, if we
                 * have one.  If we don't have room for another split chunk,
-                * add whatever space remains in this page.
+                * add whatever space remains in this page, but be careful, in
+                * the case of a large item that forced us to grow the buffer
+                * early, we may have already crossed the page-size boundary.
                 */
                r->space_avail =
                    r->split_size - WT_PAGE_HEADER_BYTE_SIZE(btree);
                if (inuse + r->space_avail > r->page_size) {
-                       WT_ASSERT(session, r->page_size >= inuse);
-                       r->space_avail = r->page_size - inuse;
+                       r->space_avail =
+                           r->page_size > inuse ? (r->page_size - inuse) : 0;
 
                        /* There are no further boundary points. */
                        r->bnd_state = SPLIT_MAX;
```

**Comment by keithbostic** [comment]
With respect to your other change (adding `r->entries == 1 || inuse + next_len < r->page_size` to the initial space-used check):

First, there's a similar check right after the `SPLIT_TRACKING_OFF` case that is roughly identical, any change we make here should probably be copied to there.

I like the pull request change better than this one for the reason you identified: if there's a small key/value pair followed by a bigger-than-page-size key/value pair, I'd rather have one larger page than two pages (one of which is tiny). Obviously, different workloads would benefit from the different policies, it's hard to argue strongly one way or the other.

**Comment by agorrod** [comment]
@keithbostic You change made sense, but it wasn't quite enough. I've pushed a variation into the branch.

For the record, this wasn't working quite how I'd expected. The sequence of calls into ```__rec_split``` is:

| Call number | r->entries | inuse | next_len | r->space_avail | r->dsk.memsize |
|---|---|---|---|---|---|
| 1 | 14 | 1173 | 3000 | 1899 | 8192 |
| 2 | 18 | 6709 | 2588 | 1483 | 8192 |

What that means is that ```__rec_split_grow``` is setting space_avail to the amount it grew the buffer (always a multiple of page size). So we are ending up with a page that contains a large item, as well as a number of smaller items after it up to the boundary.

That's reasonable, if we are going to exceed the 4k configured page size, I guess we might as well fill as close to 8k as we can. It just wasn't how I thought it worked.

File: src/reconcile/rec_write.c [hunk 2]
```diff
@@ -2649,7 +2653,7 @@ __rec_split_fixup(WT_SESSION_IMPL *session, WT_RECONCILE *r)
 	 * WT_PAGE_HEADER header onto the scratch buffer, most of the header
 	 * information remains unchanged between the pages.
 	 */
-	WT_RET(__wt_scr_alloc(session, r->page_size, &tmp));
+	WT_RET(__wt_scr_alloc(session, r->dsk.memsize, &tmp));
```
Comment (agorrod) [inline src/reconcile/rec_write.c]: @keithbostic I'm not entirely comfortable that r->dsk.memsize is the right thing to use here - though other callers of ```__rec_split_write``` pass ```r->dsk``` directly, so I guess it's the safe solution.

**Comment by keithbostic** [comment]
lgtm, merging.

**Pull request merged**